Tutorials, notebooks, merged in changes from develop (#845)
great-expectations/great_expectations: docs/tutorials/validate_data.rst

Validate Data
==============

Expectations describe data assets. Data assets are composed of batches. Validation checks expectations against a batch of data.

Validation = checking if a batch of data from a data asset X conforms to all expectations in expectation suite Y. Expectation suite Y is a collection of expectations that you created that specify what a valid batch of data asset X should look like.

To run validation you need:

* to specify which **data asset** the batch is from
* to specify an **expectation suite** to validate against
* to obtain the **batch** of data to be validated

> 

Reviewer: ```suggestion
To run validation you need a **batch** of data. To get a **batch** of data you need:
* to specify which **data asset** the batch is from
* to specify an **expectation suite** to validate against
```

Author: +1

This tutorial will explain each of these objects, show how to obtain them, execute validation and view its result.

Video
------

If you prefer videos to written tutorials, `James <https://github.com/jcampbell>`_ (one of the original core contributors) walks you through this turorial in a `video on YouTube <https://greatexpectations.io/videos/getting_started/integrate_expectations>`_.

0. Open Jupyter Notebook
------------------------

This tutorial assumes that:

* you ran ``great_expectations init`` and went through the steps covered in the previous tutorial: :ref:`tutorial_init`.
* your current directory is the root of the project where you ran ``great_expectations init``

> The dataset is a folder with CSV files containing National Provider Identifier (NPI) data that are processed with pandas.

Reviewer: Same comment. How do I get the data.

.. content-tabs::

    .. tab-container:: tab0
        :title: pandas

        .. code-block:: bash

            jupyter notebook great_expectations/notebooks/pandas/validation_playground.ipynb

    .. tab-container:: tab1
        :title: pyspark

        .. code-block:: bash

            jupyter notebook great_expectations/notebooks/spark/validation_playground.ipynb

    .. tab-container:: tab2
        :title: SQLAlchemy

        .. code-block:: bash

            jupyter notebook great_expectations/notebooks/sql/validation_playground.ipynb

1. Get a DataContext Object
---------------------------

A DataContext represents a Great Expectations project. It organizes datasources, notification settings, data documentation sites, and storage and access for expectation suites and validation results.
The DataContext is configured via a yml file stored in a directory called great_expectations;
the configuration file as well as managed expectation suites should be stored in version control.

Obtaining a DataContext object gets us access to these resources after the object reads its

> configuration file.

Reviewer: Same comments from the create tutorial.

Author: +1

    .. tab-container:: tab2
        :title: SQLAlchemy

        A SQLAlchemy datasource generates data assets from tables, views and query results.

        * If the data resided in a table (or view) in a database, it would be accessible as a data asset with the name of that table (or view).
        * If the data did not reside in one table ``npidata`` and, instead, the example pipeline ran an SQL query that fetched the data (probably from multiple tables), the result set of that query would be accessible as a data asset. The name of this data asset would be up to us (e.g., "npidata" or "npidata_query").

Great Expectations' ``jupyter_ux`` module has a convenience method that lists all data assets and expectation suites known to a Data Context:

.. code-block:: python

    great_expectations.jupyter_ux.list_available_data_asset_names(context)

Here is the output of this method when executed in our example project:

.. image:: ../images/list_data_assets.png
    :width: 600px

> ``npidata`` is the short name of the data asset. Full names of data assets in a DataContext consist of three parts, for example: ``data__dir/default/npidata``. You don't need to know (yet) how the namespace is managed and the exact meaning of each part. The :ref:`data_context` article describes this in detail.

Reviewer: Same comments from the create tutorial.

3. Pick a data asset and and expectation suite
----------------------------------------------

The previous section showed how to list all data assets and expectation suites in a project.

In this section you choose a data asset name from this list.

> You set The ``normalize_data_asset_name`` method converts the short name of a data asset to a full name:

Reviewer: Remove two sentences and fix a partial thought.
```suggestion
The ``normalize_data_asset_name`` method converts the short name of a data asset to a full name:
```

Author: +1

.. code-block:: python

    data_asset_name = "npidata"
    normalized_data_asset_name = context.normalize_data_asset_name(data_asset_name)
    normalized_data_asset_name

Choose the expectation suite you will validate the batch against:

.. code-block:: python

    expectation_suite_name = "warning"

3.a. If you don't have an expectation suite, let's create a simple one
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

If you don't have an expectation suite for this data asset, the notebook's next cell will create a suite of very basic expectations, so that you have some expectations to play with. The expectation suite will have `expect_column_to_exist` expectations for each column.

If you created an expectation suite for this data asset, you can skip executing the next cell (if you execute it, it will do nothing).

4. Load a batch of data to validate
-----------------------------------

Expectations describe data assets. Data assets are composed of batches. Validation checks expectations against a batch of data.

>  For example, a batch could be the most recent day of log data. For a database table, a batch could be the data in that table at a particular time.

Reviewer: remove the tab
```suggestion
For example, a batch could be the most recent day of log data. For a database table, a batch could be the data in that table at a particular time.
```

5. Set a Run Id
---------------

A ``run_id`` links together validations of different data assets, making it possible to track "runs" of a pipeline and
follow data assets as they are transformed, joined, annotated, enriched, or evaluated. The run id can be any string;
by default, Great Expectations will use an ISO 8601-formatted UTC datetime string.

The Great Expectations DataContext object uses the run id to determine when :ref:`evaluation_parameters` should be
linked between data assets. After each validation completes, the DataContext identifies and stores any validation
results that are referenced by other data asset expectation suites managed by the context. Then, when a batch of data
is validated against one of those expectation suites, *with the same run id*, the context will automatically insert
the relevant parameter from the validation result. For example, if a batch of the ``node_2`` data_asset expects the
number of unique values in its ``id`` column to equal the number of distinct values in the ``id`` column from
``node_1``, we need to provide the same run_id to ensure that parameter can be passed between validations.

> See :ref:`data_context_evaluation_parameter_store` for more information.

Reviewer: This is waaaay to complicated for this tutorial.

Reviewer: Remove this section entirely.

Reviewer: It might also be good to note that your pipeline runner probably has a run id that can be inserted here to make smoother integration.

Author: Agreed - done.

The default ``run_id`` generated by Great Expectations is built using the following code:

.. code-block:: python

    run_id = datetime.datetime.utcnow().isoformat().replace(":", "") + "Z"

6. Validate the batch
---------------------

Validation evaluates our expectations against the given batch and produces a report that describes observed values and
any places where expectations are not met. To validate the batch of data call the :meth:`~great_expectations.\
data_asset.data_asset.DataAsset.validate` method on the batch:

.. code-block:: python

  validation_result = batch.validate(run_id=run_id)

In a data pipeline we may take specific actions based on the the result of the validation.

> A common pattern is to check ``validation_result``'s ``success`` key (``True`` if the batch meets all the expectations in the expectation suite), and stop or issue a warning in the code in case of failure:

Reviewer: I completely love this.

  else:
    logger.warning("This file is not a valid batch of {0:s}".format(str(data_asset_name)))

The ``validation_result`` object has detailed information about every expectation in the suite that was used to validate the batch: whether the batch met the expectation and even more details if it did not. You can read more about the result object's structure here: :ref:`validation_result`.

You can print this object out:

.. code-block:: python

    print(json.dumps(validation_result, indent=4))

Here is what a part of this object looks like:

.. image:: ../images/validation_playground_result_json.png
    :width: 500px

> Don't worry - this blob of JSON is meant for machines. :ref:`data_docs` provide an HTML view of both expectation suites and validation results that is more suitable for humans. You will see how to get to them in the next sections.

Reviewer: ```suggestion
Don't panic! This blob of JSON is meant for machines. :ref:`data_docs` are an compiled HTML view of both expectation suites and validation results that is far more suitable for humans. You will see how easy it is to build them in the next sections.
```

7. Validation Operators
-----------------------

> The `validate` method evaluates one batch of data against one expectation suite and returns a dictionary of validation results. This is sufficient when you explore your data and get to know Great Expectations.

Reviewer: ```suggestion
The ``validate()`` method evaluates one batch of data against one expectation suite and returns a dictionary of validation results. This is sufficient when you explore your data and get to know Great Expectations.
```

> When deploying Great Expectations in a real data pipeline, you will typically discover additional needs:

Reviewer: ```suggestion

When deploying Great Expectations in a real data pipeline, you will typically discover these additional needs:
```

* validating a group of batches that are logically related
* validating a batch against several expectation suites

> * doing something with the validation results (e.g., saving them for a later review, sending notifications in case of failures, etc.).

Reviewer: ```suggestion
* Validating a group of batches that are logically related (eg Did all my salesforce integrations work last night?).
* Validating a batch against several expectation suites (eg Does my nightly clickstream event job have any **critical** failures I need to deal with asap or **warnings** I should investigate later?).
* Doing something with the validation results (e.g., saving them for a later review, sending notifications in case of failures, etc.).
```

Author: +1

Validation Operators provide a convenient abstraction for both bundling the validation of multiple expectation suites and the actions that should be taken after the validation. See the
:ref:`validation_operators_and_actions` for more information.

> An instance of ``action_list_operator`` operator is configured in the default ``great_expectations.yml`` configuration file. ActionListValidationOperator validates each batch in the list that is passed as `assets_to_validate` argument to its `run` method against the expectation suite included within that batch and then invokes a list of configured actions on every validation result.

Reviewer: ```suggestion
An instance of ``action_list_operator`` operator is configured in the default ``great_expectations.yml`` configuration file. ``ActionListValidationOperator`` validates each batch in the list that is passed as ``assets_to_validate`` argument to its ``run`` method against the expectation suite included within that batch and then invokes a list of configured actions on every validation result.
```

> Below is the operator's configuration:

Reviewer: ```suggestion
Below is the operator's configuration snippet in the ``great_expectations.yml`` file:
```

            module_name: great_expectations.render.renderer.slack_renderer
            class_name: SlackRenderer

We will show how to use the two most commonly used actions that are available to this operator:

> Save Validation Results

Reviewer: I'm questioning whether demonstrating S3 is in scope for this tutorial. However this should not be addressed now.

  validation_notification_slack_webhook: https://slack.com/your_webhook_url

8. View the Validation Results in Data Docs
-------------------------------------------

Data Docs compiles raw Great Expectations objects including Expectations and Validations into structured documents such as HTML documentation. By default the HTML website is hosted on your local filesystem. When you are working in a team, the website can be hosted in the cloud (e.g., on S3) and serve as the shared source of truth for the team working on the data pipeline.

Read more about the capabilities and configuration of Data Docs here: :ref:`data_docs`.

One of the actions executed by the validation operator in the previous section rendered the validation result as HTML and added this page to the Data Docs site.

You can open the page programmatically and examine the result:

.. code-block:: python

    context.open_data_docs()

> 

Reviewer: Same suggestions from the create tutorial. Probably worth a copy paste
